In [89]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OrdinalEncoder
from keras.models import Sequential
from keras.layers import Dense


In [90]:
def load_dataset(filename):
    data = pd.read_csv(filename, header=None)
    dataset = data.values
    #split to input x and output y
    x = dataset[:,:-1]
    y = dataset[:,-1]

    x = x.astype(str) #format values as string
    y = y.reshape((len(y),1)) #reshape target to 2d array

    return x, y
    

In [91]:
x, y = load_dataset('breast-cancer.csv')
#split intro train and test sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.33, random_state=1)

In [92]:
print('Train', x_train.shape, y_train.shape)
print('Test', x_test.shape, y_test.shape)

Train (191, 9) (191, 1)
Test (95, 9) (95, 1)


We have 191 samples for training and 95 for testing.

In [93]:
def input_prep(x_train, x_test):
    oe = sklpre.OrdinalEncoder()
    oe.fit(x_train)
    x_train_enc = oe.transform(x_train)
    x_test_enc = oe.transform(x_test)
    return x_train_enc, x_test_enc

In [94]:
def output_prep(y_train, y_test):
    le = sklpre.LabelEncoder()
    le.fit(y_train)
    y_train_enc = le.transform(y_train)
    y_test_enc = le.transform(y_test)
    return y_train_enc, y_test_enc

In [95]:
x_train_enc, x_test_enc = input_prep(x_train, x_test)
y_train_enc, y_test_enc = output_prep(y_train, y_test)

d:\Users\Obsidian\Documents\GitHub\MLpractice\.venv\Lib\site-packages\sklearn\preprocessing\_label.py:99: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
d:\Users\Obsidian\Documents\GitHub\MLpractice\.venv\Lib\site-packages\sklearn\preprocessing\_label.py:134: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, dtype=self.classes_.dtype, warn=True)
d:\Users\Obsidian\Documents\GitHub\MLpractice\.venv\Lib\site-packages\sklearn\preprocessing\_label.py:134: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, dtype=self.classes_.dtype, warn=True)


Neural Network Model:

In [96]:
# define the model
model = Sequential()
model.add(Dense(10, input_dim=x_train_enc.shape[1], activation='relu', kernel_initializer='he_normal'))
model.add(Dense(1, activation='sigmoid'))
# compile the keras model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
# fit the keras model on the dataset
model.fit(x_train_enc, y_train_enc, epochs=100, batch_size=16, verbose=2)
# evaluate the keras model
_, accuracy = model.evaluate(x_test_enc, y_test_enc, verbose=0)
print('Accuracy: %.2f' % (accuracy*100))

Epoch 1/100
12/12 - 1s - loss: 1.0130 - accuracy: 0.7277 - 529ms/epoch - 44ms/step
Epoch 2/100
12/12 - 0s - loss: 0.9226 - accuracy: 0.7277 - 14ms/epoch - 1ms/step
Epoch 3/100
12/12 - 0s - loss: 0.8406 - accuracy: 0.7277 - 15ms/epoch - 1ms/step
Epoch 4/100
12/12 - 0s - loss: 0.7820 - accuracy: 0.7173 - 13ms/epoch - 1ms/step
Epoch 5/100
12/12 - 0s - loss: 0.7338 - accuracy: 0.7225 - 12ms/epoch - 1ms/step
Epoch 6/100
12/12 - 0s - loss: 0.7079 - accuracy: 0.7225 - 13ms/epoch - 1ms/step
Epoch 7/100
12/12 - 0s - loss: 0.6870 - accuracy: 0.7120 - 13ms/epoch - 1ms/step
Epoch 8/100
12/12 - 0s - loss: 0.6786 - accuracy: 0.7016 - 14ms/epoch - 1ms/step
Epoch 9/100
12/12 - 0s - loss: 0.6695 - accuracy: 0.7016 - 15ms/epoch - 1ms/step
Epoch 10/100
12/12 - 0s - loss: 0.6638 - accuracy: 0.7120 - 14ms/epoch - 1ms/step
Epoch 11/100
12/12 - 0s - loss: 0.6588 - accuracy: 0.7016 - 14ms/epoch - 1ms/step
Epoch 12/100
12/12 - 0s - loss: 0.6533 - accuracy: 0.6963 - 15ms/epoch - 1ms/step
Epoch 13/100
12/12 - 0s